### PR TITLE
CC-1222: update cpp version, and add vcpkg support for dependency resolution

### PR DESCRIPTION
**Pull request by ryan-gang**
Update C++ version support to 23.
Add vcpkg for dependency resolution.
Build process also includes repository-aware caching now.

### CODE REVIEW

File: compiled_starters/cpp/CMakeLists.txt [hunk 1]
```diff
@@ -1,8 +1,11 @@
 cmake_minimum_required(VERSION 3.13)
+
 project(git-starter-cpp)
-set(CMAKE_CXX_STANDARD 20) # Enable the C++20 standard
+
+set(CMAKE_CXX_STANDARD 23) # Enable the C++23 standard
 
 file(GLOB_RECURSE SOURCE_FILES src/*.cpp)
```
Comment (rohitpaulk) [inline compiled_starters/cpp/CMakeLists.txt:7]: Let's see if this requires `src/*.hpp`, saw it in the bittorrent one and we do mention `.hpp` files in our language guide for C++

File: compiled_starters/cpp/your_git.sh [hunk 2]
```diff
@@ -6,6 +6,13 @@
...
-make > /dev/null
-exec ./server "$@"
\ No newline at end of file
+# vcpkg & cmake are required. 
+cmake -B build -S . -DCMAKE_TOOLCHAIN_FILE=${VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake
+cmake --build ./build
+
+# Check if CODECRAFTERS_SUBMISSION_DIR is set, otherwise set it to "."
+if [ -z "${CODECRAFTERS_SUBMISSION_DIR}" ]; then
+    CODECRAFTERS_SUBMISSION_DIR="."
```
Comment (rohitpaulk) [inline compiled_starters/cpp/your_git.sh:15]: @ryan-gang can't we use `$(dirname $0)` for this? On user's machines, `CODECRAFTERS_SUBMISSION_DIR` won't be set.

Comment (rohitpaulk) [reply compiled_starters/cpp/your_git.sh:15]: Our instructions for the Git course recommend that users setup an alias like `mygit`, so this should work with that

File: compiled_starters/cpp/CMakeLists.txt (hunk 1 above)
Comment (ryan-gang) [reply compiled_starters/cpp/CMakeLists.txt:7]: For the bittorrent starter we ship with a couple of .hpp files, that's why we add to the CMakeLists, here as there are no .hpp currently present, does it make sense to add them here ?
I feel adding this to the docs might be more helpful. That would also be useful across all courses.

Comment (rohitpaulk) [reply compiled_starters/cpp/CMakeLists.txt:7]: Dunno, seems like a very commonly used thing, when I was working on C++ for Redis I ran into this too if I'm not wrong. Can we just add this to all courses? This isn't an "edge-case", it's something almost everyone will run into

**Review by rohitpaulk — APPROVED** [review]

**Pull request merged**